File: python/src/simple_semantic/canonical_json.py
"""Canonical JSON encoding — SPEC.md §7.

Two implementations in two languages must emit the same bytes for the same
content, so the encoding is pinned rather than delegated to whatever a JSON
library does by default.

Python's ``json.dumps`` already satisfies the rules given the right flags, so
this module is mostly a validator plus a thin wrapper. The Kotlin side
hand-writes the equivalent encoder.
"""

from __future__ import annotations

import json
from typing import Any

from .errors import MetaValueError

_INT64_MIN = -(2**63)
_INT64_MAX = 2**63 - 1


def validate_meta(meta: dict[str, Any], _path: str = "meta") -> None:
    """Reject anything the sibling implementation could not reproduce byte-for-byte.

    Checked at write time rather than read time: writing a file the other
    implementation cannot reproduce is the failure we are preventing, and by
    read time it has already happened.
    """
    for key, value in meta.items():
        if not isinstance(key, str):
            raise MetaValueError(_path, key)
        _validate_value(value, f"{_path}.{key}")


def _validate_value(value: Any, path: str) -> None:
    if value is None or isinstance(value, str):
        return
    # bool before int: bool is a subclass of int in Python and would otherwise
    # be range-checked as an integer and serialized as one.
    if isinstance(value, bool):
        return
    if isinstance(value, int):
        if not _INT64_MIN <= value <= _INT64_MAX:
            raise MetaValueError(path, value)
        return
    if isinstance(value, list):
        for i, item in enumerate(value):
            _validate_value(item, f"{path}[{i}]")
        return
    if isinstance(value, dict):
        for key, item in value.items():
            if not isinstance(key, str):
                raise MetaValueError(path, key)
            _validate_value(item, f"{path}.{key}")
        return
    raise MetaValueError(path, value)


def encode(obj: dict[str, Any], *, sort_keys: bool = False) -> bytes:
    """Encode one object to canonical UTF-8 JSON with no trailing newline.

    ``sort_keys`` is False for objects whose key order SPEC.md fixes
    (manifest, document lines) and True for user-supplied ``meta``, whose keys
    sort ascending by code point.

    Note that Python sorts ``str`` by code point, which is what §7 requires —
    a UTF-16 code-unit sort (the JVM's natural ``String`` ordering) differs
    above the BMP, so the Kotlin encoder sorts explicitly by code point.
    """
    return json.dumps(
        obj,
        ensure_ascii=False,
        separators=(",", ":"),
        sort_keys=sort_keys,
        allow_nan=False,
    ).encode("utf-8")


def encode_string(value: str) -> bytes:
    """A single JSON string literal, escaped per §7 rule 3."""
    return json.dumps(value, ensure_ascii=False).encode("utf-8")


def encode_document(doc_id: str, text: str, meta: dict[str, Any], content_hash: str) -> bytes:
    """One line of docs.jsonl, without the terminating LF. SPEC.md §4.

    Assembled piece by piece rather than from a dict so that the outer key
    order (id, text, meta, hash) is fixed by this code while ``meta``'s
    user-supplied keys are sorted independently.
    """
    validate_meta(meta)
    return b"".join(
        [
            b'{"id":',
            encode_string(doc_id),
            b',"text":',
            encode_string(text),
            b',"meta":',
            encode(meta, sort_keys=True),
            b',"hash":',
            encode_string(content_hash),
            b"}",
        ]
    )


def decode_line(line: bytes) -> dict[str, Any]:
    parsed: dict[str, Any] = json.loads(line.decode("utf-8"))
    return parsed

File: kotlin/core/src/main/kotlin/dev/simplesemantic/CanonicalJson.kt
package dev.simplesemantic

/**
 * The canonical JSON encoder — SPEC.md §7.
 *
 * Hand-written because no JVM JSON library guarantees all four rules by default:
 * separator style, key ordering, escape selection, and literal non-ASCII output,
 * and byte-identity between the two implementations depends on all four.
 *
 * Decoding has no such constraint and is declarative — see Wire.kt.
 */
public object CanonicalJson {

    /**
     * Encode an object whose key order is already correct. Pass a
     * [LinkedHashMap] and insertion order is output order.
     */
    public fun encodeObject(entries: Map<String, Any?>, sortKeys: Boolean = false): String {
        val builder = StringBuilder()
        writeObject(builder, entries, sortKeys, "")
        return builder.toString()
    }

    /** One line of docs.jsonl, without the terminating LF. SPEC.md §4. */
    public fun encodeDocument(
        id: String,
        text: String,
        meta: Map<String, Any?>,
        hash: String,
    ): String {
        validateMeta(meta)
        val builder = StringBuilder()
        builder.append("{\"id\":")
        writeString(builder, id)
        builder.append(",\"text\":")
        writeString(builder, text)
        builder.append(",\"meta\":")
        // meta is user-supplied, so its keys sort; the four outer keys do not.
        writeObject(builder, meta, sortKeys = true, path = "meta")
        builder.append(",\"hash\":")
        writeString(builder, hash)
        builder.append('}')
        return builder.toString()
    }

    /** A single JSON string literal, escaped per §7 rule 3. */
    public fun encodeString(value: String): String {
        val builder = StringBuilder()
        writeString(builder, value)
        return builder.toString()
    }

    /** Reject anything the other implementation could not reproduce byte-for-byte. */
    public fun validateMeta(meta: Map<String, Any?>) {
        for ((key, value) in meta) {
            validateValue(value, "meta.$key")
        }
    }

    private fun validateValue(value: Any?, path: String) {
        when (value) {
            null, is String, is Boolean -> return
            is Byte, is Short, is Int, is Long -> return
            is List<*> -> value.forEachIndexed { i, item -> validateValue(item, "$path[$i]") }
            is Map<*, *> -> for ((key, item) in value) {
                if (key !is String) throw MetaValueException(path, key)
                validateValue(item, "$path.$key")
            }
            else -> throw MetaValueException(path, value)
        }
    }

    private fun writeValue(builder: StringBuilder, value: Any?, path: String) {
        when (value) {
            null -> builder.append("null")
            is String -> writeString(builder, value)
            is Boolean -> builder.append(if (value) "true" else "false")
            is Byte, is Short, is Int, is Long -> builder.append(value.toString())
            is List<*> -> {
                builder.append('[')
                value.forEachIndexed { i, item ->
                    if (i > 0) builder.append(',')
                    writeValue(builder, item, "$path[$i]")
                }
                builder.append(']')
            }
            is Map<*, *> -> {
                @Suppress("UNCHECKED_CAST")
                writeObject(builder, value as Map<String, Any?>, true, path)
            }
            else -> throw MetaValueException(path, value)
        }
    }

    private fun writeObject(
        builder: StringBuilder,
        entries: Map<String, Any?>,
        sortKeys: Boolean,
        path: String,
    ) {
        val keys = if (sortKeys) entries.keys.sortedWith(CODE_POINT_ORDER) else entries.keys.toList()
        builder.append('{')
        keys.forEachIndexed { i, key ->
            if (i > 0) builder.append(',')
            writeString(builder, key)
            builder.append(':')
            writeValue(builder, entries[key], if (path.isEmpty()) key else "$path.$key")
        }
        builder.append('}')
    }

    /**
     * Ascending by Unicode code point, per SPEC.md §7 rule 2.
     *
     * Not [String.compareTo], which orders by UTF-16 code unit and so sorts
     * astral-plane characters below U+E000..U+FFFF. Comparing UTF-8 bytes is
     * equivalent to comparing code points.
     */
    private val CODE_POINT_ORDER: Comparator<String> = Comparator { left, right ->
        val a = left.toByteArray(Charsets.UTF_8)
        val b = right.toByteArray(Charsets.UTF_8)
        var i = 0
        while (i < a.size && i < b.size) {
            val diff = (a[i].toInt() and 0xFF) - (b[i].toInt() and 0xFF)
            if (diff != 0) return@Comparator diff
            i++
        }
        a.size - b.size
    }

    private fun writeString(builder: StringBuilder, value: String) {
        builder.append('"')
        for (ch in value) {
            when {
                ch == '"' -> builder.append("\\\"")
                ch == '\\' -> builder.append("\\\\")
                ch == '\b' -> builder.append("\\b")
                ch == '\u000C' -> builder.append("\\f")
                ch == '\n' -> builder.append("\\n")
                ch == '\r' -> builder.append("\\r")
                ch == '\t' -> builder.append("\\t")
                ch < ' ' -> builder.append("\\u").append(HEX[ch.code])
                // Everything else literal, including all non-ASCII. '/' is never escaped.
                else -> builder.append(ch)
            }
        }
        builder.append('"')
    }

    /** Lowercase hex, per SPEC.md §7 rule 3. Only the C0 controls need it. */
    private val HEX: Array<String> = Array(0x20) { String.format("%04x", it) }
}
